test/test_python_gen.py

```diff
@@ -1198,6 +1217,55 @@ def test_struct_decoding_with_optional_struct(self):
         self.assertEqual(s2.f1.f1, 'hello')
         self.assertEqual(s2.f1.f2, 3)
 
+    def test_struct_equality_with_object(self):
+        """Should not throw an error when comparing with object.
+
+        Object is a superclass of Struct, but it should not be considered for equality.
+        """
+        s = self.decode(self.sv.Struct(self.ns.S), json.dumps({'f': 'F'}))
+        self.assertFalse(s == object())
+
+    def test_struct_equality_with_value(self):
+        s = self.decode(self.sv.Struct(self.ns.S), json.dumps({'f': 'F'}))
+        s_equal = self.decode(self.sv.Struct(self.ns.S), json.dumps({'f': 'F'}))
+        s_unequal = self.decode(self.sv.Struct(self.ns.S), json.dumps({'f': 'G'}))
+        self.assertEqual(s, s_equal)
+        self.assertEqual(hash(s), hash(s_equal))
+        self.assertNotEqual(s, s_unequal)
+        self.assertNotEqual(hash(s), hash(s_unequal))
+
+    def test_struct_equality_with_different_types(self):
+        """Structs of different types that do not have an inheritance relationship are not considered
+        equal to each other."""
+        s = self.decode(self.sv.Struct(self.ns.S), json.dumps({'f': 'F'}))
+        t = self.decode(self.sv.Struct(self.ns.T), json.dumps({'f': 'F'}))
+        self.assertNotEqual(s, t)
+        # They still hash to the same value, since they have the same fields, but this is
+        # fine since we don't expect to use a large number of structs as dict keys.
+        self.assertEqual(hash(s), hash(t))
+
+        # S_extend and S_extend2 are indirectly related because they both extend S, but they do not
+        # have a direct line of inheritance to each other.
+        s_extend = self.decode(self.sv.Struct(self.ns.SExtend), json.dumps({'f': 'F', 'g': 'G'}))
+        s_extend2 = self.decode(self.sv.Struct(self.ns.SExtend2), json.dumps({'f': 'F', 'g': 'G'}))
+        self.assertNotEqual(s_extend, s_extend2)
+        # They still hash to the same value, since they have the same fields, but this is
+        # fine since we don't expect to use a large number of structs as dict keys.
+        self.assertEqual(hash(s_extend), hash(s_extend2))
+
+    def test_extended_struct_equality(self):
+        """Structs which subclass each other are considered equal to each other."""
+        s = self.decode(self.sv.Struct(self.ns.S), json.dumps({'f': 'F'}))
+        s_extend = self.decode(self.sv.Struct(self.ns.SExtend), json.dumps({'f': 'F', 'g': 'G'}))
+        s_extend_extend = self.decode(self.sv.Struct(self.ns.SExtendExtend),
+                                      json.dumps({'f': 'F', 'g': 'G', 'h': 'H'}))
+        self.assertEqual(s, s_extend)
+        self.assertNotEqual(hash(s), hash(s_extend))
```

Comment:
Two points on `__eq__` and `__hash__`
1. The hash of an object does not change across the object's lifetime
2.  a == b must imply hash(a) == hash(b). The reverse is not necessarily true

From the Python Docs [here](https://docs.python.org/3.7/reference/datamodel.html#object.__hash__)
>  If a class defines mutable objects and implements an `__eq__()` method, it should not implement `__hash__()`, since the implementation of hashable collections requires that a key’s hash value is immutable (if the object’s hash value changes, it will be in the wrong hash bucket).

Although the Python Docs do not call out having `True` for `__eq__` implies same hash, the default behavior for user defined class shows Python follows this invariant
>User-defined classes have `__eq__()` and `__hash__()` methods by default; with them, all objects compare unequal (except with themselves) and `x.__hash__()` returns an appropriate value such that `x == y` implies both that x is y and `hash(x) == hash(y)`.

See the articles [here](https://www.asmeurer.com/blog/posts/what-happens-when-you-mess-with-hashing-in-python/) and [here](https://hynek.me/articles/hashes-and-equality/) for detailed examples of the two points